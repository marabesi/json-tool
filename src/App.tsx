import { useState } from 'react'
import JsonEditor from './components/JsonEditor'
import Label from './components/Label';
import Formatter from './core/formatter'

function App() {
  const [originalJson, setOriginalResult] = useState('');
  const [result, setResult] = useState('');

  const onJsonChange = async (value: string) => {
    const format = new Formatter(value);
    const result = await format.format();

    setOriginalResult(value);
    setResult(result);
  }

  return (
    <div className="h-screen bg-gray-500 p-5">
      <div className="flex h-3/6">
        <div className="w-3/6 flex flex-col h-full m-1">
          <Label>place your json here</Label>
          <JsonEditor
            input={originalJson}
            onChange={eventValue => onJsonChange(eventValue.value)}
            data-testid="json"
            className="original"
          />
        </div>
        <div className="w-3/6 flex flex-col h-full m-1">
          <Label>result</Label>
          <JsonEditor
            input={result}
            className="result"
          />
        </div>
      </div>
    </div>
  );
}

export default App;
