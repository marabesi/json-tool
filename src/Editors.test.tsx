import { render, screen, act, waitFor, fireEvent, within } from '@testing-library/react';
import App from './App';
import userEvent from '@testing-library/user-event';

function grabCurrentEditor(container: HTMLElement): HTMLElement {
  const editor = container.querySelector('[data-testid="json"] .cm-content');
  if (!editor) {
    throw new Error('Could not find editor');
  }
  return editor as HTMLElement;
}

describe('Editors', () => {
  beforeEach(() => {
    document.createRange = () => {
      const range = new Range();

      range.getBoundingClientRect = jest.fn();

      range.getClientRects = () => {
        return {
          item: () => null,
          length: 0,
          [Symbol.iterator]: jest.fn()
        };
      };

      return range;
    };
  });

  it.each([
    ['{{}', '{}'],
    ['{{"a": "b"}', '{"a": "b"}'],
  ])('place %s text in the editor and receive %s', async (input, expected) => {
    const { container } = render(<App />);

    const editor = grabCurrentEditor(container);

    await act(async () => {
      await userEvent.type(editor, input);
    });

    const result = screen.getByTestId('result');

    expect(result.nodeValue).toMatchSnapshot(expected);
  });

  it('should keep content in the editor when navigating away', async () => {
    const { container, getByTestId, getByText } = render(<App/>);

    const editor = grabCurrentEditor(container);
    const json = '{{"random_json":"123"}';

    await act(async () => {
      await userEvent.type(editor, json);
    });

    const rawEditor = screen.getByTestId('raw-json');

    await waitFor(() => {
      expect(rawEditor).toHaveValue('{"random_json":"123"}');
    });

    fireEvent.click(getByTestId('settings'));

    await waitFor(() => {
      expect(getByText('Settings')).toBeInTheDocument();
    });

    fireEvent.click(getByTestId('to-home'));

    await waitFor(() => {
      expect(screen.getByTestId('raw-json')).toHaveValue('{"random_json":"123"}');
    });
  });

  it('should render search element in the json editor', async () => {
    const { getByTestId } = render(<App/>);

    fireEvent.click(getByTestId('search-json'));

    await waitFor(() => expect(within(getByTestId('json')).getByText('×')).toBeInTheDocument());
  });

  it('should render search element in the result editor', async () => {
    const { getByTestId } = render(<App/>);

    fireEvent.click(getByTestId('search-result'));

    await waitFor(() => expect(within(getByTestId('result')).getByText('×')).toBeInTheDocument());
  });
});
