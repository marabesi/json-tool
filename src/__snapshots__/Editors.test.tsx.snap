// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editors place {{"a": "b"} text in the editor and receive {"a": "b"}: {"a": "b"} 1`] = `null`;

exports[`Editors place {{} text in the editor and receive {}: {} 1`] = `null`;
